tests/method_wrappers/test_data_reducer.py

```diff
@@ -0,0 +1,73 @@
+import math
+from httomo.method_wrappers import make_method_wrapper
+from httomo.method_wrappers.datareducer import DatareducerWrapper
+from httomo.runner.auxiliary_data import AuxiliaryData
+from httomo.runner.dataset import DataSetBlock
+from ..testing_utils import make_mock_repo
+
+
+import numpy as np
+from mpi4py import MPI
+from pytest_mock import MockerFixture
+
+
+def test_datareducer(mocker: MockerFixture):
+    class FakeModule:
+        def data_reducer(x):
+            return 2 * x
+
+    mocker.patch("importlib.import_module", return_value=FakeModule)
+    wrp = make_method_wrapper(
+        make_mock_repo(mocker),
+        "mocked_module_path.morph",
+        "data_reducer",
+        MPI.COMM_WORLD,
+    )
+    assert isinstance(wrp, DatareducerWrapper)
+
+    GLOBAL_SHAPE = (10, 20, 30)
+    aux_data = AuxiliaryData(
+        angles=np.linspace(0, math.pi, GLOBAL_SHAPE[0], dtype=np.float32),
+        darks=4 * np.ones((2, GLOBAL_SHAPE[1], GLOBAL_SHAPE[2]), dtype=np.float32),
+        flats=3 * np.ones((3, GLOBAL_SHAPE[1], GLOBAL_SHAPE[2]), dtype=np.float32),
+    )
+    data = np.ones(GLOBAL_SHAPE, dtype=np.float32)
+
+    block1 = DataSetBlock(
+        data[0:2, :, :],
+        aux_data=aux_data,
+        slicing_dim=0,
+        block_start=0,
+        chunk_start=0,
+        chunk_shape=GLOBAL_SHAPE,
+        global_shape=GLOBAL_SHAPE,
+    )
+    newblock1 = wrp.execute(block1)
+
+    # we double flats and darks only, so we expect them to be twice the input
+    np.testing.assert_array_equal(newblock1.data, 1)
+    np.testing.assert_array_equal(newblock1.flats, 6)
+    np.testing.assert_array_equal(newblock1.darks, 8)
+
+    # it should only update the darks/flats once
```

Comment:
What this comment says is important, that the darks/flats should only be updated when the first block is processed, but I think this isn't being checked in the test?

I wonder if it may be useful to have another test for the data reducer wrapper which asserts this somehow? Ie, another test that somehow explicitly checks that when you process the second block, the darks/flats have definitely not been reassigned by the wrapper's `execute()` method?

Reply:
There's also the option of creating a mock of the `AuxiliaryData` object that is passed to the `DataSetBlock` constructor, to then "spy" on if the darks/flats setter get called only once, like what is done in one of the task runner tests to check that side outputs are only updated once https://github.com/DiamondLightSource/httomo/blob/0bff94bf70a75c9fadbfd14cae24e62a6c66d9fd/tests/runner/test_task_runner.py#L221

Reply:
Fair point indeed. I've added a couple of things to cover it, but feel free to change if you think there is a better way. 

- I'm checking that `_flats_darks_processed` variable set to `True` after executing the first block
- For the second block I change the auxiliary data for flats and darks, so if on the second block we do process f/d  again, then those updated arrays will be used instead of the initial ones. The assertion to initial f/d arrays proves that we ignore the updated ones but not executing the condition. Hope that this logic make sense?

Reply:
Thanks for the update!

Sorry, I'm admittedly a bit confused about the second auxiliary data object :sweat_smile:  I'm trying to make sure I understand how the extra `AuxiliaryData` object created in the test is being used, can I walk through what I _think_ I understand, and then you can correct me where I've got something wrong:
- `block1` is created with auxiliary data such that its darks is an array of 4's, and its flats is an array of 3's
- when `block1` is processed, the darks and flats should be doubled, so `newblock1` should have its darks as an array of 8's, and its flats as an array of 6's
- a new auxiliary data object is created with darks all 8's and flats all 6's, which are the same _values_ as the processed darks/flats in `newblock1`, but different array objects (ie, the values in the arrays are equal, but the array objects are not equal)
- `block2` is given this new auxiliary data that has darks all 8's and flats all 6's, so it has darks/flats which are the same values as the darks/flats from `newblock1`
- when `block2` is processed, its darks and flats should _not_ be doubled, so darks should stay as all 8's and flats should stay as all 6's
- at the end, the darks/flats of `newblock2` are asserted to be 8's for darks and 6's for flats

One thing that I might be misunderstanding is, because `block2` is given the same values for darks/flats as the doubled darks/flats from `newblock1`, how are we able to distinguish between:
- `newblock2` has 8's for darks and 6's for flats because it was given the new auxiliary data object
- `newblock2` has 8's for darks and 6's for flats because it has the updated darks/flats from `newblock1`

After all, `np.testing.assert_array_equal()` is only checking equality of the values in the arrays, it's not checking the array objects are the same object?

Is the idea that the darks/flats given to `block2` _haven't_ doubled from 8 to 18 and from 6 to 12 respectively?

Reply:
Something else that I'm not sure about is that different blocks should never be given different `AuxiliaryData` objects, a fundamental part of `AuxiliaryData`'s purpose was that every block should hold a reference to the _same_ `AuxiliaryData` object.

Creating two auxiliary data objects to give to two different blocks isn't something that should happen during pipeline execution, so it doesn't feel ideal to be testing a method wrapper under that condition, a method wrapper would always be given blocks which have a reference to the same `AuxiliaryData` object.

Reply:
Regarding the spy that was added, sorry, I think I didn't explain what I meant clearly enough when I suggested it as an option!

What I was meaning was that we could spy on the darks/flats setter on the `AuxiliaryData` object, to verify that the setters only get called once (ie, to check that the processing of the second block doesn't call the setters).

As it is now, the spy is looking at the wrapper's `execute()` method, which I don't think is doing any verification of the darks/flats being modified or not. What the spy is being used right now for is simply confirming that the wrapper's `execute()` method is called once after the test runs the `execute()` method, which isn't doing much -> we know that the `execute()` method has been called once because the test explicitly does this. I think what we're more interested in is spying on the call to any function that modifies the darks/flats (because checking how many times a function that modifies darks/flats is called will tell us how many times the darks/flats have been modified).

Does that make a bit more sense?

Reply:
I've made a few changes to the data reducer wrapper test in [494d2bf](https://github.com/DiamondLightSource/httomo/pull/219/commits/494d2bf04dccfa601469ed70e82df63124f857f6) to better illustrate what I was trying to describe.

(We don't have to keep that commit by any means, if it doesn't look useful we can revert locally and then force push to undo it, I just thought it was easier to show what I was meaning in code rather then describing it only in words :slightly_smiling_face:)